File: yage/src/yage/Time.cpp
#include "Time.hpp"

namespace yage {

Time::Time(Time::Seconds fixedDelta)
        : fixedDelta_(fixedDelta),
          fixedDeltaRaw_(fixedDelta.count()) {}

void Time::start() {
    lastTimeSeen_ = Clock::now();
    startTime_ = lastTimeSeen_;
}

void Time::accumulate() {
    auto currentTime = Clock::now();
    auto difference = currentTime - lastTimeSeen_;
    delta_ = Seconds(difference);
    accumulatedTime_ += delta_;
    lastTimeSeen_ = currentTime;
}

bool Time::consumeFixedDelta() {
    if (accumulatedTime_ > fixedDelta_) {
        accumulatedTime_ -= fixedDelta_;
        fixedStepCount_++;
        return true;
    }
    return false;
}
}

File: rust_games/hindsight/src/time.rs
pub struct FixedTimeLoop {
    accumulated_time: f32,
    delta_time: f32,
}

impl FixedTimeLoop {
    pub fn new(delta_time: f32) -> Self {
        Self {
            accumulated_time: 0.0,
            delta_time,
        }
    }

    pub fn accumulate(&mut self, time: f32) {
        self.accumulated_time += time;
    }

    pub fn step(&mut self) -> bool {
        if self.accumulated_time >= self.delta_time {
            self.accumulated_time -= self.delta_time;
            true
        } else {
            false
        }
    }
}
